tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +121,82 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob writing procedure works as follows:
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The caller sends a number of WriteBlobRequests, providing the blob data
+//    in an ordered sequence of chunks.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC.  Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single chunk suffices, so the single request is both
+//    the first and the last.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * The caller may choose to verify each response before sending the next
+//    request, or it may continue sending requests optimistically.
+//
+// Note that the WriteBlobRequest does not mirror the nested structure of
+// WriteScalarRequest, because we only ever send one blob at a time.
+
+// Metadata included with only the first chunk of the blob write stream.
```

Comment:
Potentially no action required: it’s also true that metadata is only
required on the first blob sequence in a time series. Up to you whether
this should be noted/clarified.

Reply:
Refactoring in the works, pending offline discussion.

Reply:
I think the updated design re CreateBlobSequence vs WriteBlob fully clarifies this.